pep-0690.rst

```diff
@@ -166,35 +183,212 @@ will be enough to trigger the import of ``spam.py``::
+overhead on normal non-lazy dictionaries (just one check to see if the
+dictionary has a ``*_LAZY`` lookup kind).
+
+The eagerness of imports within ``try`` / ``except`` / ``with`` blocks or within
+class or function bodies is handled in the compiler via a new
+``EAGER_IMPORT_NAME`` opcode that always imports eagerly. Top-level imports use
+``IMPORT_NAME``, which may be lazy or eager depending on ``-L`` and/or
+``importlib.enable_lazy_imports()``.
+
+
+Exceptions
+----------
+
+Exceptions that occur during a lazy import bubble up and erase the
+partially-constructed module(s) from ``sys.modules``, just as exceptions during
+normal import do.
+
+Since errors raised during a lazy import will occur later (wherever the imported
+name is first referenced) than they would if the import were eager, it is
+possible that they could be accidentally caught by exception handlers that
+didn't expect the import to be running within their ``try`` block, leading to
+confusion. To reduce the potential for this confusion, exceptions raised in the
+course of executing a lazy import will be replaced by a ``LazyImportError``
+exception (a subclass of ``ImportError``), with ``__cause__`` set to the
+original exception.
+
+The ``LazyImportError`` will have source location metadata attached pointing the
+user to the original import statement, to ease debuggability of errors from lazy
+imports. (It won't have a full traceback to the original import location; this
+is too expensive to preserve for all lazy imports, and it's not clear that it
+provides significant value over simply knowing the location of the import
+statement.)
+
+Only ``Exception`` are replaced in this way, not ``BaseException``.
+``BaseException`` are for "system-exiting" exceptions like ``KeyboardInterrupt``
+or ``SystemExit``; these are normally not caught, and if they are caught, it is
+less likely to be specific to a certain bit of code that was expected to raise
+them, and more likely that the goal is to catch them whatever their origin.
+
+
+Debugging
+---------
+
+Debug logging from ``python -v`` will include logging whenever an import
+statement has been encountered but execution of the import will be deferred.
 
 Python's ``-X importtime`` feature for profiling import costs adapts naturally
 to lazy imports; the profiled time is the time spent actually importing.
 
+Although lazy import objects are never visible to Python code, in some debugging
+cases it may be useful to check from Python code whether the value at a given
+key in a given dictionary is a lazy import object, without triggering its
+resolution. For this purpose, ``importlib.is_lazy_import()`` can be used::
+
+    from importlib import is_lazy_import
+
+    import foo
+
+    is_lazy_import(globals(), "foo")
+
+    foo
+
+    is_lazy_import(globals(), "foo")
+
+In this example, if lazy imports have been enabled the first call to
+``is_lazy_import`` will return ``True`` and the second will return ``False``.
```

Comment:
We don't yet have all of this implemented (``is_lazy_import`` was feedback from the discussion thread, not something we needed internally, and we aren't adding new features until we get the 3.12 port done), so I can't use an _actual_ interpreter session yet :) Barring that I'm not sure what the advantage is; it adds prompt cruft but I'm not sure it makes anything clearer?

Reply:
Okay. Something about the code block confuses me, since if this is intended as code to put in a module and execute, then the extra blank lines are confusing and I also don't know what the return values are.  That's why I thought a (mock) interactive interpreter session would look nicer.  Just a thought and not substantial to the purpose of the PEP.